### PR TITLE
Generate valid OpenAPI v3.0.0 documents

### PR DESCRIPTION
**Pull request by SKalt**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
Previously, `./api/openapi-spec/swagger.json` was an invalid OpenAPI 2.x document. This prevented [`openapi-generator`][openapi-generator] from generating clients for Harvester. 

**Solution:**

- I added custom post-processing code to fix some mistakes that `kube-openapi`'s codegen didn't catch.
  - 2f37fecb fix: move schema definition out of path name and into param definition
  - 48157742 fix: prevent [`uniqueItems`](https://json-schema.org/draft/2020-12/json-schema-validation#name-uniqueitems) from being set on non-array JSON schemas
  - a8358343 fix: add [security schemes](https://spec.openapis.org/oas/v3.0.3#security-scheme-object) to OpenAPI generation
  - 20e7e8db fix: include string pattern validation in output OpenAPI v3 schema

Unfortunately, in an attempt to get those fixes to work, I had pulled in a year's worth of `kube-openapi` changes, which are the majority of the changes in this PR. If the dependency upgrades break CI, I might be able to revert some of them.

**Related Issue:**
#5211

**Test plan:**

```sh
set -eu
go run pkg/genswagger/main.go
swagger-cli validate ./api/openapi-spec/swagger.json | sed 's/^/# /g`
# ./api/openapi-spec/swagger.json is valid

target_dir=$TEMPDIR/go-sdk
mkdir -p $target_dir
openapi-generator-cli generate \
  --generator-name go \
  -i ./api/openapi-spec/swagger.json \
  -o $target_dir >/tmp/generation.log 2>&1
```

---

Suggested reading order (~300 lines of diff)
- `./go.mod`
- `./pkg/genswagger/rest/webservices.go`
- `./pkg/genswagger/main.go`

[kube-openapi]: https://github.com/kubernetes/kube-openapi
[openapi-generator]: https://openapi-generator.tech/

### CODE REVIEW

**Review by FrankYang0529 — APPROVED** [review]
LGTM. Thanks for the contribution.

File: pkg/genswagger/main.go [hunk 1]
```diff
@@ -121,20 +317,46 @@ func createConfig() *common.Config {
...
+			ensureRequiredPropsUnique(swagger)
+			return swagger, nil
+		},
 	}
 }
 
-func addSummaryForMethods(swagger *spec.Swagger) {
-	t := reflect.TypeOf(new(spec.Operation))
+func addSummaryForMethods(swagger *spec3.OpenAPI) {
+	// t := reflect.TypeOf(new(spec3.Operation))
```
Comment (Yu-Jack) [inline pkg/genswagger/main.go]: you could remove this comment if it's not needed.

File: pkg/genswagger/main.go [hunk 2]
```diff
@@ -76,16 +78,210 @@ func main() {
...
+			// find all the schemas in the spec
+			for _, param := range op.Parameters {
+				if param != nil {
+					if schema := param.Schema; schema != nil {
+						if !schema.Type.Contains("array") {
+							schema.UniqueItems = false
+						}
+					}
+				}
+			}
```
Comment (Yu-Jack) [inline pkg/genswagger/main.go]: just keep it consistent

```suggestion
			for _, param := range op.Parameters {
				if param == nil {
					continue
				}
				if schema := param.Schema; schema != nil {
					if !schema.Type.Contains("array") {
						schema.UniqueItems = false
					}
				}
			}
```

File: pkg/genswagger/main.go [hunk 3]
```diff
@@ -76,16 +78,210 @@ func main() {
...
+	}
+	return
+}
+
+var kindToParamSourceName = map[int]string{
+	restful.PathParameterKind:   "path",
+	restful.QueryParameterKind:  "query",
+	restful.BodyParameterKind:   "body",
+	restful.HeaderParameterKind: "header",
+}
```
Comment (Yu-Jack) [inline pkg/genswagger/main.go]: you could move it to the top

File: go.mod [hunk 4]
```diff
@@ -22,7 +22,7 @@ replace (
 	k8s.io/client-go => k8s.io/client-go v0.26.13
 	k8s.io/cloud-provider => k8s.io/cloud-provider v0.26.13
 	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.26.13
-	k8s.io/code-generator => k8s.io/code-generator v0.26.13
+	k8s.io/code-generator => k8s.io/code-generator v0.26.13 // TODO: update -> v0.29.2; see https://github.com/kubernetes/code-generator/compare/v0.26.13...v0.29.2
```
Comment (Yu-Jack) [inline go.mod]: I think we could remove this TODO?

**Review by Yu-Jack — COMMENTED** [review]
Thank for your contributions! There are just some NITs.

**Comment by SKalt** [comment]
@Yu-Jack nits addressed! LMK what you think.

**Review by Yu-Jack — APPROVED** [review]
LGTM, thank for your contributions!

File: vendor/modules.txt [hunk 5]
```diff
@@ -1121,10 +1121,6 @@ golang.org/x/crypto/pbkdf2
 golang.org/x/crypto/scrypt
 golang.org/x/crypto/ssh
 golang.org/x/crypto/ssh/internal/bcrypt_pbkdf
-# golang.org/x/exp v0.0.0-20230725012225-302865e7556b
```
Comment (SKalt) [inline vendor/modules.txt]: From https://github.com/harvester/harvester/commit/7af889bb8ce9ed8cb8ce901012a3940916cec5b4#diff-b318e1e6bbcc59892bb6b4298564b051c6f6d5d848878d14aab856509ff0625c

File: go.mod [hunk 6]
```diff
@@ -95,7 +95,6 @@ require (
 	github.com/urfave/cli v1.22.13
 	go.uber.org/multierr v1.11.0
 	golang.org/x/crypto v0.21.0
-	golang.org/x/exp v0.0.0-20230725012225-302865e7556b
```
Comment (SKalt) [inline go.mod]: from https://github.com/harvester/harvester/commit/7af889bb8ce9ed8cb8ce901012a3940916cec5b4#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6

**Review by jillian-maroket — APPROVED** [review]
@SKalt Thank you for contributing.

@bk201 I'll do a separate string review when the schedule permits.

**Comment by bk201** [comment]
@SKalt I can't merge the PR due to a conflict, can you help rebase the PR? Thanks.

**Comment by SKalt** [comment]
Hey @bk201, what branch and commit is this PR conflicting with? I'm not seeing any conflicts with `master`. If you'd like me to squash/fixup the commit log, LMK and I can do that too.

**Comment by bk201** [comment]
Hi, it should be a conflict with the master. I am seeing this:
截圖 2024-04-02 晚上11 02 41

Can you fetch the harvester remote and rebase on the master again? And it would be nice if you could squash commits too.

**Comment by SKalt** [comment]
Oof -- I'm not seeing that at all! 
Screenshot 2024-04-02 at 11 21 28 AM
Do you want to hop on a call to see if we can screenshare and work this out?

**Comment by brandboat** [comment]
@mergify backport v1.3

**Pull request merged**